README.md
     __  __  _ __   ____  
    /\ \/\ \/\`'__\/',__\ 
    \ \ \_\ \ \ \//\__, `\
     \ \____/\ \_\\/\____/
      \/___/  \/_/ \/___/... Universal Reddit Scraper 

![GitHub top language](https://img.shields.io/github/languages/top/JosephLai241/URS?logo=Python)
[![PRAW Version](https://img.shields.io/badge/PRAW-7.1.4-red?logo=Reddit)][PRAW]
[![Build Status](https://img.shields.io/travis/JosephLai241/URS?logo=Travis)][Travis CI Build Status]
[![Codecov](https://img.shields.io/codecov/c/gh/JosephLai241/URS?logo=Codecov)][Codecov]
[![GitHub release (latest by date)](https://img.shields.io/github/v/release/JosephLai241/URS)][Releases]
![Total lines](https://img.shields.io/tokei/lines/github/JosephLai241/URS)
![License](https://img.shields.io/github/license/JosephLai241/URS)

<p align="center"> 
    <img 
        alt="Help Screenshot"
        src="https://i.imgur.com/KP1A0mZ.png">
</p>

# Table of Contents

* [Contact](#contact)
* [Introduction](#introduction)
* [URS Overview](#urs-overview)
    + [Export File Format](#export-file-format)
    + [Export Directory Structure](#export-directory-structure)
    + [Scrape Speeds](#scrape-speeds)
    + [Scraping Reddit via PRAW](#scraping-reddit-via-praw)
        * [Getting Started](#getting-started)
        * [Rate Limits](#rate-limits)
        * [Table of All Subreddit, Redditor, and Submission Comments Attributes](#a-table-of-all-subreddit-redditor-and-submission-comments-attributes)
        * [Subreddits](#subreddits)
        * [Redditors](#redditors)
        * [Submission Comments](#submission-comments)
    + [Analytical Tools](#analytical-tools)
        * [Target Fields](#target-fields)
        * [File Names](#file-names)
        * [Generating Word Frequencies](#generating-word-frequencies)
        * [Generating Wordclouds](#generating-wordclouds)
    + [Exporting](#exporting)
* [Contributing](#contributing)
    + [Before Making Pull or Feature Requests](#before-making-pull-or-feature-requests)
    + [Building on Top of URS](#building-on-top-of-urs)
    + [Making Pull or Feature Requests](#making-pull-or-feature-requests)
* [Contributors](#contributors)
* [Derivative Projects](#derivative-projects)
* Supplemental Documents
    + [How to get Reddit API Credentials for PRAW][How to get Reddit API Credentials for PRAW]
    + [Error Messages][Error Messages]
    + [2-Factor Authentication][2-Factor Authentication]
    + [Releases/Changelog][Releases/Changelog]
    + [Some Linux Tips][Some Linux Tips]

# Contact

Whether you are using URS for enterprise or personal use, I am very interested in hearing about your use case and how it has helped you achieve a goal. 

Additionally, please send me an email if you would like to [contribute](#contributing), have questions, or want to share something you have built on top of it. 

You can send me an email or leave a note by clicking on either of these badges. I look forward to hearing from you!

[![Email](https://img.shields.io/badge/Email-urs__project%40protonmail.com-informational?logo=ProtonMail)][URS Project Email]
[![Say Thanks!](https://img.shields.io/badge/Say%20Thanks-!-blue)][Say Thanks!]

# Introduction

This is a comprehensive Reddit scraping tool that integrates multiple features:

* Scrape Reddit via [`PRAW`][PRAW] (the official Python Reddit API Wrapper)
    + Scrape Subreddits
    + Scrape Redditors
    + Scrape submission comments
* Analytical tools for scraped data
    + Get frequencies for words that are found in submission titles, bodies, and/or comments
    + Generate a wordcloud from scrape results
    
Run `pip install -r requirements.txt` to get all project dependencies. 

You will need your own Reddit account and API credentials for PRAW. See the [Getting Started](#getting-started) section for more information. 

***NOTE:*** `PRAW` is currently supported on Python 3.6+.

# URS Overview

## Export File Format

**All files except for those generated by the wordcloud tool are exported to JSON by default**. Wordcloud files are exported to PNG by default. URS supports exporting to CSV as well, but JSON is the more versatile option. See the [Exporting](#exporting) section for more information.

## Export Directory Structure

All exported files are saved within the `scrapes` directory and stored in a sub-directory labeled with the date. Many more sub-directories may be created in the date directory. Sub-directories are only created when its respective tool is run. For example, if you only use the Subreddit scraper, only the `subreddits` directory is created.

The `subreddits`, `redditors`, or `comments` directories are created when you run each scraper.

The `analytics` directory is created when you run any of the analytical tools. Within it, the `frequencies` or `wordclouds` directories are created when you run each tool. See the [Analytical Tools](#analytical-tools) section for more information.

This is the [samples][Samples]' directory structure generated by the [tree command][tree].

```
scrapes/
└── 02-24-2021
    ├── analytics
    │   ├── frequencies
    │   │   ├── askreddit-hot-100-results.csv
    │   │   └── askreddit-hot-100-results.json
    │   └── wordclouds
    │       ├── cscareerquestions-search-'job'-past-year.png
    │       └── What do you think about adults that don’t eat ve---RAW.png
    ├── comments
    │   ├── What do you think about adults that don’t eat ve---50-results.json
    │   └── What do you think about adults that don’t eat ve---RAW.json
    ├── redditors
    │   └── spez-5-results.json
    ├── subreddits
    │   ├── askreddit-hot-100-results.csv
    │   ├── askreddit-hot-100-results.json
    │   ├── cscareerquestions-search-'job'-past-year.json
    │   └── wallstreetbets-top-5-results-past-month-rules.json
    └── urs.log
```

## Scrape Speeds

Scrape speed is determined by a couple things:

* The number of results returned for Subreddit or Redditor scraping
* The submission's popularity (total number of comments) for submission comments scraping
* Your internet connection speed

## Scraping Reddit via PRAW

### Getting Started

It is very quick and easy to get Reddit API credentials. Refer to [my guide][How to get Reddit API Credentials for PRAW] to get your credentials, then update the `API` dictionary located in `Credentials.py`

### Rate Limits

Yes, PRAW has rate limits. These limits are proportional to how much karma you have accumulated - the higher the karma, the higher the rate limit. This has been implemented to mitigate spammers and bots that utilize PRAW.

Rate limit information for your account is displayed in a small table underneath the successful login message each time you run any of the PRAW scrapers. I have also added a `--check` flag if you want to quickly view this information.

URS will display an error message as well as the rate limit reset date if you have used all your available requests.

There are a couple ways to go about solving issues with rate limits:

* Scrape intermittently
* Use an account with high karma to get your PRAW credentials
* Scrape less results per run

Available requests are refilled if you use the PRAW scrapers intermittently, which might be a good solution for bypassing rate limit issues. This can be especially helpful if you have automated URS and are not looking at the output on each run.

---

### A Table of All Subreddit, Redditor, and Submission Comments Attributes

These attributes are included in each scrape. 

| Subreddits    | Redditors                      | Submission Comments |
|---------------|--------------------------------|---------------------|
| Title         | Name                           | Parent ID           |
| Flair         | Fullname                       | Comment ID          |
| Date Created  | ID                             | Author              |
| Upvotes       | Date Created                   | Date Created        |
| Upvote Ratio  | Comment Karma                  | Upvotes             |
| ID            | Link Karma                     | Text                |
| Is Locked?    | Is Employee?                   | Edited?             |
| NSFW?         | Is Friend?                     | Is Submitter?       |
| Is Spoiler?   | Is Mod?                        | Stickied?           |
| Stickied?     | Is Gold?                       |                     |
| URL           | \*Submissions                  |                     |
| Comment Count | \*Comments                     |                     |
| Text          | \*Hot                          |                     |
| &nbsp;        | \*New                          |                     |
| &nbsp;        | \*Controversial                |                     |
| &nbsp;        | \*Top                          |                     |
| &nbsp;        | \*Upvoted (may be forbidden)   |                     |
| &nbsp;        | \*Downvoted (may be forbidden) |                     |
| &nbsp;        | \*Gilded                       |                     |
| &nbsp;        | \*Gildings (may be forbidden)  |                     |
| &nbsp;        | \*Hidden (may be forbidden)    |                     |
| &nbsp;        | \*Saved (may be forbidden)     |                     |

\*Includes additional attributes; see [Redditors](#redditors) section for more information. 

---

### Subreddits

![Subreddit Demo GIF][Subreddit Demo]

\*This GIF is uncut.

**Usage:** `$ ./Urs.py -r SUBREDDIT (H|N|C|T|R|S) N_RESULTS_OR_KEYWORDS` 

**Supported export formats:** JSON and CSV. To export to CSV, include the `--csv` flag.

You can specify Subreddits, the submission category, and how many results are returned from each scrape. I have also added a search option where you can search for keywords within a Subreddit.

These are the submission categories:

* Hot
* New
* Controversial
* Top
* Rising
* Search

The file names for all categories except for Search will follow this format: 

`"[SUBREDDIT]-[POST_CATEGORY]-[N_RESULTS]-result(s).[FILE_FORMAT]"`

If you searched for keywords, file names will follow this format:

`"[SUBREDDIT]-Search-'[KEYWORDS]'.[FILE_FORMAT]"` 

### Time Filters

Time filters may be applied to some categories. Here is a table of the categories on which you can apply a time filter as well as the valid time filters.

| Categories    | Time Filters  | 
|---------------|---------------|
| Controversial | All (default) |
| Top           | Day           |
| Search        | Hour          |
| &nbsp;        | Month         | 
| &nbsp;        | Week          |
| &nbsp;        | Year          |

Specify the time filter after the number of results returned or keywords you want to search for.

**Usage:** `$ ./Urs.py -r SUBREDDIT (C|T|S) N_RESULTS_OR_KEYWORDS OPTIONAL_TIME_FILTER`

If no time filter is specified, the default time filter `all` is applied. The Subreddit settings table will display `None` for categories that do not offer the additional time filter option.

If you specified a time filter, `-past-[TIME_FILTER]` will be appended to the file name before the file format like so: 

`"[SUBREDDIT]-[POST_CATEGORY]-[N_RESULTS]-result(s)-past-[TIME_FILTER].[FILE_FORMAT]"` 

Or if you searched for keywords:

`"[SUBREDDIT]-Search-'[KEYWORDS]'-past-[TIME_FILTER].[FILE_FORMAT]"`

### Subreddit Rules

You can also include the Subreddit's rules and post requirements in your scrape data by including the `--rules` flag. **This only works when exporting to JSON.**

If rules are included in your file, `-rules` will be appended to the end of the file name.

### Bypassing the Final Settings Check

After submitting the arguments and Reddit validation, URS will display a table of Subreddit scraping settings as a final check before executing. You can include the `-y` flag to bypass this and immediately scrape.

Exported files will be saved to the `subreddits` directory.

***NOTE:*** Up to 100 results are returned if you search for keywords within a Subreddit. You will not be able to specify how many results to keep.

---

### Redditors

![Redditor Demo GIF][Redditor Demo]

\*This GIF has been cut for demonstration purposes.

**Usage:** `$ ./Urs.py -u REDDITOR N_RESULTS` 

**Supported export formats:** JSON.

You can also scrape Redditor profiles and specify how many results are returned.

Some Redditor attributes are sorted differently. Here is a table of how each is sorted.

| Attribute Name | Sorted By/Time Filter                       |
|----------------|---------------------------------------------|
| Comments       | Sorted By: New                              |
| Controversial  | Time Filter: All                            |
| Gilded         | Sorted By: New                              |
| Hot            | Determined by other Redditors' interactions |
| New            | Sorted By: New                              |
| Submissions    | Sorted By: New                              |
| Top            | Time Filter: All                            |

Of these Redditor attributes, the following will include additional attributes:

| Submissions, Hot, New, Controversial, Top, Upvoted, Downvoted, Gilded, Gildings, Hidden, and Saved | Comments                                     |
|----------------------------------------------------------------------------------------------------|----------------------------------------------|
| Title                                                                                              | Date Created                                 |
| Date Created                                                                                       | Score                                        |
| Upvotes                                                                                            | Text                                         |
| Upvote Ratio                                                                                       | Parent ID                                    |
| ID                                                                                                 | Link ID                                      |
| NSFW?                                                                                              | Edited?                                      |
| Text                                                                                               | Stickied?                                    |
| &nbsp;                                                                                             | Replying to (title of submission or comment) |
| &nbsp;                                                                                             | In Subreddit (Subreddit name)                |

The file names will follow this format: 

`"[USERNAME]-[N_RESULTS]-result(s).json"` 

Exported files will be saved to the `redditors` directory.

***NOTE:*** If you are not allowed to access a Redditor's lists, PRAW will raise a 403 HTTP Forbidden exception and the program will just append `"FORBIDDEN"` underneath that section in the exported file. 

***NOTE:*** The number of results returned are applied to all attributes. I have not implemented code to allow users to specify different number of results returned for individual attributes. 

---

### Submission Comments

![Structured Comments Demo GIF][Structured Comments Demo]
![Raw Comments Demo GIF][Raw Comments Demo]

\*These GIFs have been cut for demonstration purposes.

**Usage:** `$ ./Urs.py -c SUBMISSION_URL N_RESULTS` 

**Supported export formats:** JSON.

You can also scrape comments from submissions and specify the number of results returned. 

Comments are sorted by "Best", which is the default sorting option when you visit a submission.

**There are two ways you can scrape comments: structured or raw.** This is determined by the number you pass into `N_RESULTS`:

| Scrape Type | N_RESULTS      |
|-------------|----------------|
| Structured  | N_RESULTS >= 1 |
| Raw         | N_RESULTS = 0  |

Structured scrapes resemble comment threads on Reddit and will include down to third-level comment replies. 

Raw scrapes do not resemble comment threads, but returns all comments on a submission in level order: all top-level comments are listed first, followed by all second-level comments, then third, etc.

Of all scrapers included in this program, this usually takes the longest to execute. PRAW returns submission comments in level order, which means scrape speeds are proportional to the submission's popularity.

The file names will follow this format: 

`"[POST_TITLE]-[N_RESULTS]-result(s).json"` 

Exported files will be saved to the `comments` directory.

***NOTE:*** You cannot specify the number of raw comments returned. The program with scrape all comments from the submission. 

## Analytical Tools

This suite of tools can be used *after* scraping data from Reddit. Both of these tools analyze the frequencies of words found in submission titles and bodies, or comments within JSON scrape data.

There are a few ways you can quickly get the correct filepath to the scrape file:

* Drag and drop the file into the terminal. 
* Partially type the path and rely on tab completion support to finish the full path for you.

Running either tool will create the `analytics` directory within the date directory. **This directory is located in the same directory in which the scrape data resides**. For example, if you run the frequencies generator on February 16th for scrape data that was captured on February 14th, `analytics` will be created in the February 14th directory. Command history will still be written in the February 16th `urs.log`.

A shortened export path is displayed once URS has completed exporting the data, informing you where the file is saved within the `scrapes` directory. You can open `urs.log` to view the full path. 

The sub-directories `frequencies` or `wordclouds` are created in `analytics` depending on which tool is run.

***NOTE:*** Do not move the `scrapes` directory elsewhere if you want to use these tools. URS uses a relative path to save the generated files.

---

### Target Fields

The data varies depending on the scraper, so these tools target different fields for each type of scrape data:

| Scrape Data         | Targets                 |
|---------------------|-------------------------|
| Subreddit           | "title", "text"         |
| Redditor            | "title", "body", "text" |
| Submission Comments | "text"                  |

For Subreddit scrapes, data is pulled from the "title" and "text" fields for each submission (submission title and body).

For Redditor scrapes, data is pulled from all three fields because both submission and comment data is returned. The "title" and "body" fields are targeted for submissions, and the "text" field is targeted for comments.

For submission comments scrapes, data is only pulled from the "text" field of each comment.

---

### File Names

File names are identical to the original scrape data so that it is easier to distinguish which analytical file corresponds to which scrape.

---

### Generating Word Frequencies

![Frequencies Demo GIF][Frequencies Demo]

\*This GIF is uncut.

**Usage:** `$ ./Urs.py -f FILE_PATH` 

**Supported export formats:** JSON and CSV. To export to CSV, include the `--csv` flag.

You can generate a dictionary of word frequencies created from the words within the target fields. 

Frequencies export to JSON by default, but this tool also works well in CSV format.

Exported files will be saved to the `analytics/frequencies` directory.

---

### Generating Wordclouds

![Wordcloud Demo GIF][Wordcloud Demo]

\*This GIF is uncut.

**Usage:** `$ ./Urs.py -wc FILE_PATH`

**Supported export formats:** .eps, .jpeg, .jpg, .pdf, .png (default), .ps, .rgba, .tif, .tiff.

Taking word frequencies to the next level, you can generate wordclouds based on word frequencies. This tool is independent of the frequencies generator - you do not need to run the frequencies generator before creating a wordcloud. 

PNG is the default format, but you can also export to any of the options listed above by including the format as the second flag argument.

**Usage:** `$ ./Urs.py -wc FILE_PATH OPTIONAL_EXPORT_FORMAT` 

Exported files will be saved to the `analytics/wordclouds` directory.

### Display Wordcloud Instead of Saving

Wordclouds are saved to file by default. If you do not want to keep a file, include the `--nosave` flag to only display the wordcloud.

## Exporting

As stated before, URS supports exporting to either JSON or CSV. **JSON is the default format** - you will have to include the `--csv` flag to export to CSV.

I recommend only exporting to CSV when using:

+ The Subreddit scraper
+ The word frequencies generator

These tools are also suitable for CSV format and are optimized to do so if you want to use that format instead.

JSON is the more practical option for Redditor and submission comments scraping, which is why I have designed these scrapers to work best in this format. It is much easier to read the scrape results since Redditor scraping returns attributes that include additional submission or comment attributes. 

Comments scraping is especially easier to read in JSON format because structured exports look similar to threads on Reddit. You can process all the information pertaining to a comment much quicker compared to CSV. 

You can still export Redditor data and submission comments to CSV, but you will be disappointed with the results.

### See the [samples][Samples] for scrapes ran on February 24, 2021.

# Contributing

**See the [Contact](#contact) section for ways to reach me.**

## Before Making Pull or Feature Requests

Consider the scope of this project before submitting a pull or feature request. URS stands for Universal Reddit Scraper. Two important aspects are listed in its name - *universal* and *scraper*.

I will not approve feature or pull requests that deviate from its sole purpose. This may include scraping a specific aspect of Reddit or [adding functionality that allows you to post a comment with URS][Commenting Feature Request]. Adding either of these requests will no longer allow URS to be universal or merely a scraper. However, I am more than happy to approve requests that enhance the current scraping capabilities of URS.

## Building on Top of URS

Although I will not approve requests that deviate from the project scope, feel free to reach out if you have built something on top of URS or have made modifications to scrape something specific on Reddit. I will add your project to the [Derivative Projects](#derivative-projects) section!

## Making Pull or Feature Requests

You can suggest new features or changes by going to the [Issues tab][Issues] and fill out the Feature Request template. If there is a good reason for a new feature, I will consider adding it.

You are also more than welcome to create a pull request - adding additional features, improving runtime, or refactoring existing code. If it is approved, I will merge the pull request into the master branch and credit you for contributing to this project.

# Contributors


docs/Releases.md
# Releases

## v3.2.0 - February 25, 2021

### Added

* User Interface
    + Analytical tools
        * Word frequencies generator.
        * Wordcloud generator.
* Source code
    + CLI
        * Flags
            + `-e` - Display additional example usage.
            + `--check` - Runs a quick check for PRAW credentials and displays the rate limit table after validation.
            + `--rules` - Include the Subreddit's rules in the scrape data (for JSON only). This data is included in the `subreddit_rules` field.
            + `-f` - Word frequencies generator.
            + `-wc` - Wordcloud generator.
            + `--nosave` - Only display the wordcloud; do not save to file.
        * Added additional verbose feedback if invalid arguments are given.
    + Log decorators
        * Added new decorator to log individual argument errors.
        * Added new decorator to log when no Reddit objects are left to scrape after failing validation check.
        * Added new decorator to log when an invalid file is passed into the analytical tools.
        * Added new decorator to log when the `scrapes` directory is missing, which would cause the new `make_analytics_directory()` method in `DirInit.py` to fail. 
            + This decorator is also defined in the same file to avoid a circular import error.
    + ASCII art
        * Added new art for the word frequencies and wordcloud generators.
        * Added new error art displayed when a problem arises while exporting data.
        * Added new error art displayed when Reddit object validation is completed and there are no objects left to scrape.
        * Added new error art displayed when an invalid file is passed into the analytical tools.
* `README` 
    + Added new Contact section and moved contact badges into it.
        + Apparently it was not obvious enough in previous versions since users did not send emails to the address specifically created for URS-related inquiries.
    + Added new sections for the analytical tools.
    + Updated demo GIFs
        * Moved all GIFs to a separate branch to avoid unnecessary clones.
        * Hosting static images on Imgur.
* Tests
    + Added additional tests for analytical tools.

### Changed

* User interface
    + JSON is now the default export option. `--csv` flag is required to export to CSV instead.
    + Improved JSON structure.
        * PRAW scraping export structure:
            + Scrape details are now included at the top of each exported file in the `scrape_details` field.
                * Subreddit scrapes - Includes `subreddit`, `category`, `n_results_or_keywords`, and `time_filter`.
                * Redditor scrapes - Includes `redditor` and `n_results`.
                * Submission comments scrapes - Includes `submission_title`, `n_results`, and `submission_url`.
            + Scrape data is now stored in the `data` field.
                * Subreddit scrapes - `data` is a list containing submission objects.
                * Redditor scrapes - `data` is an object containing additional nested dictionaries: 
                    + `information` - a dictionary denoting Redditor metadata, 
                    + `interactions` - a dictionary denoting Redditor interactions (submissions and/or comments). Each interaction follows the Subreddit scrapes structure.
                * Submission comments scrapes - `data` is an list containing additional nested dictionaries.
                    + Raw comments contains dictionaries of `comment_id: SUBMISSION_METADATA`.
                    + Structured comments follows the structure seen in raw comments, but includes an extra `replies` field in the submission metadata, holding a list of additional nested dictionaries of `comment_id: SUBMISSION_METADATA`. This pattern repeats down to third level replies.
        * Word frequencies export structure:
            + The original scrape data filepath is included in the `raw_file` field.
            + `data` is a dictionary containing `word: frequency`.
    + Log:
        * `scrapes.log` is now named `urs.log`.
        * Validation of Reddit objects is now included - invalid Reddit objects will be logged as a warning.
        * Rate limit information is now included in the log.
* Source code
    + Moved PRAW scrapers into its own package.
    + Scrape settings for the basic Subreddit scraper is now cleaned within `Basic.py`, further streamlining conditionals in `Subreddit.py` and `Export.py`.
    + Returning final scrape settings dictionary from all scrapers after execution for logging purposes, further streamlining the `LogPRAWScraper` class in `Logger.py`.
    + Passing the submission URL instead of the exception into the `not_found` list for submission comments scraping.
        * This is a part of a bug fix that is listed in the Fixed section.
    + ASCII art:
        * Modified the args error art to display specific feedback when invalid arguments are passed.
    + Upgraded from relative to absolute imports.
    + Replaced old header comments with docstring comment block.
    + Upgraded method comments to Numpy/Scipy docstring format.
* `README`
    + Moved Releases section into its own document.
    + Deleted all media from master branch.
* Tests
    + Updated absolute imports to match new directory structure.
    + Updated a few tests to match new changes made in the source code.
* Community documents
    + Updated `PULL_REQUEST_TEMPLATE`:
        * Updated section for listing changes that have been made to match new Releases syntax.
        * Wrapped New Dependencies in a code block.
    + Updated `STYLE_GUIDE`:
        * Created new rules for method comments.
    + Added `Releases`:
        * Moved Releases section from main `README` to a separate document.

### Fixed

* Source code
    + PRAW scraper settings

src/cli.py
#===============================================================================
#                       Command-line Interface Functions
#===============================================================================
import argparse
import sys

from colorama import Fore, init, Style
from . import global_vars, titles, validation

init(autoreset = True)

### Global variables
short_cat = global_vars.short_cat
s_t = global_vars.s_t

usage = "scraper.py [-h] [-r SUBREDDIT [H|N|C|T|R|S] RESULTS_OR_KEYWORDS] [-u USER RESULTS] [-c URL RESULTS] [-b] [-y] [--csv|--json]"
description = "Universal Reddit Scraper 3.0 - Scrape Subreddits, submissions, Redditors, or comments from posts"
epilog = r"""
Subreddit categories:
   H,h     selecting Hot category
   N,n     selecting New category
   C,c     selecting Controversial category
   T,t     selecting Top category
   R,r     selecting Rising category
   S,s     selecting Search category

EXAMPLES

    Get the first 10 posts in r/all in the Hot category and export to JSON:

        $ ./scraper.py -r all h 10 --json

    Search for "United States of America" in r/worldnews and export to CSV:

        $ ./scraper.py -r worldnews s "United States of America" --csv

    Scraping 50 results from u/spez's Reddit account:

        $ ./scraper.py -u spez 50 --json

    Scraping 25 comments from this r/TIFU post:

        $ ./scraper.py -c https://www.reddit.com/r/tifu/comments/a99fw9/tifu_by_buying_everyone_an_ancestrydna_kit_and/ 25 --csv

    You can scrape multiple items at once:

        $ ./scraper.py -r askreddit h 15 -u spez 25 -c https://www.reddit.com/r/tifu/comments/a99fw9/tifu_by_buying_everyone_an_ancestrydna_kit_and/ 50 --json

    You can also still use URS 1.0 (SUBREDDIT SCRAPING ONLY), but you cannot include this flag with any items besides export options:

        $ ./scraper.py -b --csv

"""

submissions_params = ["after", "aggs", "author", "before", "contest_mode", "fields", 
                        "frequency", "ids", "is_video", "locked", "metadata", 
                        "num_comments", "over_18", "q", "q:not", "score", "selftext", 
                        "selftext:not", "size", "sort", "sort_type", "spoiler", 
                        "stickied", "subreddit", "title", "title:not"]

### Get args
def parse_args():
    parser = argparse.ArgumentParser(usage = usage, \
                                    formatter_class = argparse.RawDescriptionHelpFormatter, \
                                    description = description, \
                                    epilog = epilog)

    ### Parser Subreddit, submissions, Redditor, comments, basic scraper, and skip confirmation flags
    scraper = parser.add_argument_group("Scraping options")
    scraper.add_argument("-r", "--subreddit", action = "append", nargs = 3, metavar = "", 
                            help = "specify Subreddit to scrape")
    scraper.add_argument("-s", "--submission", action = "append", nargs = "*", 
                            metavar = "", help = "search for keywords in any submission")
    scraper.add_argument("-u", "--redditor", action = "append", nargs = 2, metavar = "", 
                            help = "specify Redditor profile to scrape")
    scraper.add_argument("-c", "--comments", action = "append", nargs = 2, metavar = "", 
                            help = "specify the URL of the post to scrape comments")
    scraper.add_argument("-b", "--basic", action = "store_true", 
                            help = "initialize non-CLI Subreddit scraper")
    scraper.add_argument("-y", action = "store_true", 
                            help = "skip Subreddit options confirmation and scrape immediately")

    ### Export to CSV or JSON flags
    expt = parser.add_mutually_exclusive_group(required = True)
    expt.add_argument("--csv", action = "store_true", help = "export to CSV")
    expt.add_argument("--json", action = "store_true", help = "export to JSON")

    ### Print help message if no arguments are present
    if len(sys.argv[1:]) == 0:
        parser.print_help()
        parser.exit()

    args = parser.parse_args()
    return parser, args

### Create either Subreddit, Redditor, or posts list
def create_list(args, s_t, l_type):
    if l_type == s_t[0]:
        list = [sub[0] for sub in args.subreddit]
    elif l_type == s_t[1]:
        list = [user[0] for user in args.redditor]
    elif l_type == s_t[2]:
        list = [post[0] for post in args.comments]

    return list

### Check args and catching errors
def check_args(parser, args):
    try:
        if args.subreddit:
            for subs in args.subreddit:
                len_counter = 0
                if subs[1].upper() not in short_cat:
                    raise ValueError
                for char in short_cat:
                    if str(subs[1]).upper() == char:
                        if str(subs[1]).upper() != "S" and subs[2].isalpha():
                            raise ValueError
                        break
                    elif len_counter == len(short_cat) - 1:
                        raise ValueError
                    else:
                        len_counter += 1
        if args.submission:
            for submission in args.submission:
                for i in range(0, len(submission)):
                    param = submission[i].split("=")[0]
                    if param not in submissions_params:
                        print(Style.BRIGHT + Fore.RED + 
                            "\nAN INVALID SUBMISSION PARAMETER WAS ENTERED\n")
                        print(Style.BRIGHT + "CHOOSE FROM: %s\n" % 
                            ", ".join(submissions_params))
                        parser.exit()
        if args.redditor:
            for user in args.redditor:
                if user[1].isalpha():
                    raise ValueError
        if args.comments:
            for post in args.comments:
                if post[1].isalpha():
                    raise ValueError
    except ValueError:
        titles.e_title()
        parser.exit()

### Check if Subreddits exist and list invalid Subreddits if applicable
def confirm_subs(reddit, sub_list, parser):
    print("\nChecking if Subreddit(s) exist...")
    found, not_found = validation.existence(reddit, sub_list, parser, s_t, s_t[0])
    if not_found:
        print("\nThe following Subreddits were not found and will be skipped:")
        print("-" * 60)
        print(*not_found, sep = "\n")

    subs = [sub for sub in found]
    return subs

### Get CLI scraping settings for Subreddits, Redditors, and post comments
def get_cli_settings(reddit, args, master, s_t, s_type):
    if s_type == s_t[0]:
        for sub_n in master:
            for sub in args.subreddit:
                settings = [sub[1], sub[2]]
                if sub_n == sub[0]:
                    master[sub_n].append(settings)
    elif s_type == s_t[1]:
        for user in args.redditor:
            master[user[0]] = user[1]
    elif s_type == s_t[2]:
        for comments in args.comments:
            master[comments[0]] = comments[1]
